internal/enabler/enabler.go

```diff
@@ -0,0 +1,107 @@
+// package enabler contains a component that can be enabled and disabled dynamically
+package enabler
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/pomerium/pomerium/internal/log"
+)
+
+var errCauseEnabler = errors.New("enabler")
+
+// A Handler is a component with a RunEnabled function.
+type Handler interface {
+	RunEnabled(ctx context.Context) error
+}
+
+// HandlerFunc is a function run by the enabler.
+type HandlerFunc func(ctx context.Context) error
+
+func (f HandlerFunc) RunEnabled(ctx context.Context) error {
+	return f(ctx)
+}
+
+// An Enabler enables or disables a component dynamically.
+// When the Enabler is enabled, the Handler's RunEnabled will be called.
+// If the Enabler is subsequently disabled the context passed to RunEnabled will be canceled.
+// If the Enabler is subseqently enabled again, RunEnabled will be called again.
+// Handlers should obey the context lifetime and be tolerant of RunEnabled
+// being called multiple times. (not concurrently)
+type Enabler interface {
+	Run(ctx context.Context) error
+	Enable()
+	Disable()
+}
+
+type enabler struct {
+	name    string
+	handler Handler
+
+	mu      sync.Mutex
+	cancel  context.CancelCauseFunc
+	enabled bool
+}
+
+// New creates a new Enabler.
+func New(name string, handler Handler, enabled bool) Enabler {
+	d := &enabler{
+		name:    name,
+		handler: handler,
+		enabled: enabled,
+		cancel:  func(_ error) {},
+	}
+	return d
+}
+
+// Run calls RunEnabled if enabled, otherwise it waits until enabled.
+func (d *enabler) Run(ctx context.Context) error {
+	for {
+		err := d.runOnce(ctx)
+		if errors.Is(err, errCauseEnabler) {
+			continue
+		}
+		return err
+	}
```

Comment:
I find this loop a little confusing, but I think I'm slowly understanding. Let me see if I can summarize the behavior:

 - This loop will keep going until the underlying context is canceled for some reason unrelated to the enabled/disabled state.
 - If currently enabled, runOnce() will synchronously call RunEnabled(), blocking until that function returns.
 - If currently disabled, runOnce() will still block, waiting for a transition to the enabled state.

I think the name "runOnce" confused me a little (because it might not run anything), but I don't have any naming alternatives to offer.

Reply:
I changed the loop to avoid the continue.

When a `ctx` is involved like this I try to have the loop separate from the implementation so that the scope of the new `ctx` doesn't override the original one (a subtle bug that has bit me several times before), and also so we can use normal returns to exit early instead of continues.

I renamed the method to make it more clear and added some comments.

Basically we get the current `enabled` state, create a new cancelable context and call the cancel function in `Enable` and `Disable`. 

In the case when we're enabled, we call `RunEnabled` with the new context which will be canceled by a call to `Disable`. We then check the underlying cause. If it's due to us, we return that error, which is checked in the loop. But any other error (or nil) we want to exit to preserve the original behavior of `Run`.

In the case when we're disabled, we just wait for the context to be canceled, which can only happen if the parent context is canceled or we cancel it with a call to `Enable`. `context.Cause` should surface that error, and the loop will continue if the cause was us, or return if the parent context was canceled.

Reply:
Thanks, I like the name `runOrWaitForEnabled`.